### ISSUE TITLE
Factory and Provider Memory Leak

### ISSUE THREAD

**Issue by eygraber** [Kodein-Framework/Kodein-DI#301]
I added LeakCanary to my app, and I've been getting reports that any class that uses a factory or provider binding is leaking a reference to a destroyed Activity.

I didn't look too deeply into it, but it looks like the lambda in Factory.getFactory is capturing the BindingDI instance, which for some reason has a reference to my Activity in its context field. I'm not sure how the activity got set as the context. 

I use [this scope](https://github.com/eygraber/kodein-android-components/blob/master/kodein-androidx-components/src/main/java/com/eygraber/kodein/androidx/kodein.kt#L82) but nothing in SomethingApi, SomethingManager, nor somethingFactory use it. 

There is a class bound in an activitySingleton scope that uses SomethingManager but that's not referenced in the leak trace, it is bound after SomethingApi, and I've verified that it and activityScope has been properly destroyed. 

```
┬───
│ GC Root: System class
│
├─ android.provider.FontsContract class
│    Leaking: NO (MyApp↓ is not leaking and a class is never leaking)
│    ↓ static FontsContract.sContext
├─ co.my.app.MyApp instance
│    Leaking: NO (Application is a singleton)
│    MyApp does not wrap an activity context
│    ↓ MyApp.kodein
│                     ~~~~~~
├─ org.kodein.di.LazyKodein instance
│    Leaking: UNKNOWN
│    ↓ LazyKodein.baseKodein$delegate
│                 ~~~~~~~~~~~~~~~~~~~
├─ kotlin.SynchronizedLazyImpl instance
│    Leaking: UNKNOWN
│    ↓ SynchronizedLazyImpl._value
│                           ~~~~~~
├─ org.kodein.di.internal.KodeinImpl instance
│    Leaking: UNKNOWN
│    ↓ KodeinImpl._container
│                 ~~~~~~~~~~
├─ org.kodein.di.internal.KodeinContainerImpl instance
│    Leaking: UNKNOWN
│    ↓ KodeinContainerImpl.tree
│                          ~~~~
├─ org.kodein.di.internal.KodeinTreeImpl instance
│    Leaking: UNKNOWN
│    ↓ KodeinTreeImpl.bindings
│                     ~~~~~~~~
├─ java.util.HashMap instance
│    Leaking: UNKNOWN
│    ↓ HashMap.table
│              ~~~~~
├─ java.util.HashMap$Node[] array
│    Leaking: UNKNOWN
│    ↓ HashMap$Node[].[89]
│                     ~~~~
├─ java.util.HashMap$Node instance
│    Leaking: UNKNOWN
│    ↓ HashMap$Node.value
│                   ~~~~~
├─ java.util.ArrayList instance
│    Leaking: UNKNOWN
│    ↓ ArrayList.elementData
│                ~~~~~~~~~~~
├─ java.lang.Object[] array
│    Leaking: UNKNOWN
│    ↓ Object[].[0]
│               ~~~
├─ org.kodein.di.KodeinDefinition instance
│    Leaking: UNKNOWN
│    ↓ KodeinDefinition.binding
│                       ~~~~~~~
├─ org.kodein.di.bindings.Singleton instance
│    Leaking: UNKNOWN
│    ↓ Singleton.scope
│                ~~~~~
├─ org.kodein.di.bindings.NoScope instance
│    Leaking: UNKNOWN
│    ↓ NoScope._registry
│              ~~~~~~~~~
├─ org.kodein.di.bindings.StandardScopeRegistry instance
│    Leaking: UNKNOWN
│    ↓ StandardScopeRegistry._cache
│                            ~~~~~~
├─ java.util.concurrent.ConcurrentHashMap instance
│    Leaking: UNKNOWN
│    ↓ ConcurrentHashMap.table
│                        ~~~~~
├─ java.util.concurrent.ConcurrentHashMap$Node[] array
│    Leaking: UNKNOWN
│    ↓ ConcurrentHashMap$Node[].[0]
│                               ~~~
├─ java.util.concurrent.ConcurrentHashMap$Node instance
│    Leaking: UNKNOWN
│    ↓ ConcurrentHashMap$Node.val
│                             ~~~
├─ org.kodein.di.bindings.SingletonReference$make$1 instance
│    Leaking: UNKNOWN
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ SingletonReference$make$1.$value
│                                ~~~~~~
├─ co.my.app.api.SomethingApiImpl instance
│    Leaking: UNKNOWN
│    ↓ SomethingApiImpl.somethingManager
│                    ~~~~~~~~~~~~~
├─ co.my.app.models.something.SomethingManagerImpl instance
│    Leaking: UNKNOWN
│    ↓ SomethingManagerImpl.somethingFactory
│                        ~~~~~~~~~~~~~
├─ org.kodein.di.bindings.Factory$getFactory$1 instance
│    Leaking: UNKNOWN
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ Factory$getFactory$1.$kodein
│                           ~~~~~~~
├─ org.kodein.di.internal.BindingKodeinImpl instance
│    Leaking: UNKNOWN
│    ↓ BindingKodeinImpl.context
│                        ~~~~~~~
╰→ co.my.app.main.MainActivity instance
​     Leaking: YES (ObjectWatcher was watching this because co.myapp.main.MainActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
```

**Comment by eygraber** [comment]
This is still happening in Kodein v7. Here's the updated trace:

```
┬───
│ GC Root: System class
│
├─ android.app.ActivityThread class
│    Leaking: NO (a class is never leaking)
│    ↓ static ActivityThread.sCurrentActivityThread
│                            ~~~~~~~~~~~~~~~~~~~~~~
├─ android.app.ActivityThread instance
│    Leaking: UNKNOWN
│    ↓ ActivityThread.mServices
│                     ~~~~~~~~~
├─ android.util.ArrayMap instance
│    Leaking: UNKNOWN
│    ↓ ArrayMap.mArray
│               ~~~~~~
├─ java.lang.Object[] array
│    Leaking: UNKNOWN
│    ↓ Object[].[1]
│               ~~~
├─ co.my.app.feature_location.BackgroundLocationService instance
│    Leaking: UNKNOWN
│    BackgroundLocationService does not wrap an activity context
│    ↓ BackgroundLocationService.locationTracker$delegate
│                                ~~~~~~~~~~~~~~~~~~~~~~~~
├─ kotlin.SynchronizedLazyImpl instance
│    Leaking: UNKNOWN
│    ↓ SynchronizedLazyImpl._value
│                           ~~~~~~
├─ co.my.app.location_tracking.LocationTrackerImpl instance
│    Leaking: UNKNOWN
│    ↓ LocationTrackerImpl.locationApiClient
│                          ~~~~~~~~~~~~~~~~~
├─ co.my.app.location_tracking.GoogleLocationApiClientImpl instance
│    Leaking: UNKNOWN
│    ↓ GoogleLocationApiClientImpl.geofenceTriggeredReceiverIntentProvider
│                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
├─ org.kodein.di.DIContainer$provider$$inlined$toProvider$1 instance
│    Leaking: UNKNOWN
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ DIContainer$provider$$inlined$toProvider$1.$this_toProvider
│                                                 ~~~~~~~~~~~~~~~~
├─ org.kodein.di.bindings.Provider$getFactory$1 instance
│    Leaking: UNKNOWN
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ Provider$getFactory$1.$di
│                            ~~~
├─ org.kodein.di.internal.BindingDIImpl instance
│    Leaking: UNKNOWN
│    ↓ BindingDIImpl.context
│                    ~~~~~~~
╰→ co.my.app.main.MainActivity instance
​     Leaking: YES (ObjectWatcher was watching this because co.my.app.main.MainActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
```

**Comment by eygraber** [comment]
I attached a zip containing a repro project.

The issue is that in MainActivity.kt line 24, we bind `MainPresenter` in the `activitySingleton` scope ([`activitySingleton` source](https://github.com/eygraber/kodein-di-android-components/blob/master/kodein-di-androidx-components/src/main/java/com/eygraber/kodein/di/androidx/di.kt#L80)). That uses the `Activity` as the kodein context.

Because it is the first time we retrieve an instance of `SomeSingleton`, DI uses the call site context (i.e. the `Activity`). I'm not sure if that's working as intended or not, but it seems wrong to me (because `singleton` shouldn't be using another scope's context). Using `eagerSingleton` solves the problem, but I don't think that's a good general solution, since all singletons would then have to become eager.

My fix is to create another type of `NoArgDIBinding` that has the deferred behavior of `Singleton` and the context safe behavior of `EagerSingleton`. This can either replace the implementation of `Singleton` (if this is a bug) or provide an additional option (if this is working as intended).

```
class ContextSafeSingleton<T : Any>(
    builder: DIContainer.Builder,
    override val createdType: TypeToken<out T>,
    val creator: NoArgSimpleBindingDI<Any>.() -> T
) : NoArgDIBinding<Any, T> {

    override val contextType = TypeToken.Any

    @Volatile
    private var _instance: (() -> T)? = null
    private val _lock = Any()

    private fun getFactory(di: BindingDI<Any>): (Unit) -> T {
        return { _ ->
            synchronizedIfNull(
                lock = _lock,
                predicate = this@ContextSafeSingleton::_instance,
                ifNotNull = { it() },
                ifNull = {
                    NoArgBindingDIWrap(di).creator().also { _instance = { it } }
                }
            )
        }
    }

    /**
     * @see [DIBinding.getFactory]
     */
    override fun getFactory(di: BindingDI<Any>, key: DI.Key<Any, Unit, T>) = getFactory(di)

    override fun factoryName() = "eagerSingleton"

    init {
        val key = DI.Key(TypeToken.Any, TypeToken.Unit, createdType, null)
        synchronized(_lock) {
            builder.onReady { getFactory(BindingDIImpl(this, key, Any(), 0)).invoke(Unit) }
        }
    }

    override val copier =
        DIBinding.Copier { builder -> EagerSingleton(builder, createdType, creator) }
}

inline fun <reified T : Any> DI.Builder.contextSafeSingleton(
    noinline creator: NoArgSimpleBindingDI<Any>.() -> T
) = ContextSafeSingleton<T>(containerBuilder, generic(), creator)
```

Note that `init` here solves what I believe is a race condition in `EagerSingleton` in the event that something else gets called before `builder.onReady's` callback is invoked. If that is guaranteed to not happen, then the `synchronized` in `init` can be removed.

[kodeinissue301.zip](https://github.com/Kodein-Framework/Kodein-DI/files/4796020/kodeinissue301.zip)

**Comment by romainbsl** [comment]
You can try the bug fix on the dev channel, with:
- repository :`https://dl.bintray.com/kodein-framework/kodein-dev`
- kodein DI version: `7.1.0-master-80`

**Comment by eygraber** [comment]
The fix in 7.1.0-master-80 is still leaking (leak trace below). 

Is there any particular reason why replacing the implementation of `Singleton` with `ContextSafeSingleton` ([see above](https://github.com/Kodein-Framework/Kodein-DI/issues/301#issuecomment-645739116)) wouldn't work? In my testing it clears up all memory leaks.

```
┬───
│ GC Root: System class
│
├─ android.provider.FontsContract class
│    Leaking: NO (MyApp↓ is not leaking and a class is never leaking)
│    ↓ static FontsContract.sContext
├─ co.my.app.MyApp instance
│    Leaking: NO (Application is a singleton)
│    ↓ MyApp.di
│                     ~~
├─ org.kodein.di.LazyDI instance
│    Leaking: UNKNOWN
│    ↓ LazyDI.baseDI$delegate
│             ~~~~~~~~~~~~~~~
├─ kotlin.SynchronizedLazyImpl instance
│    Leaking: UNKNOWN
│    ↓ SynchronizedLazyImpl._value
│                           ~~~~~~
├─ org.kodein.di.internal.DIImpl instance
│    Leaking: UNKNOWN
│    ↓ DIImpl._container
│             ~~~~~~~~~~
├─ org.kodein.di.internal.DIContainerImpl instance
│    Leaking: UNKNOWN
│    ↓ DIContainerImpl.tree
│                      ~~~~
├─ org.kodein.di.internal.DITreeImpl instance
│    Leaking: UNKNOWN
│    ↓ DITreeImpl.bindings
│                 ~~~~~~~~
├─ java.util.HashMap instance
│    Leaking: UNKNOWN
│    ↓ HashMap.table
│              ~~~~~
├─ java.util.HashMap$Node[] array
│    Leaking: UNKNOWN
│    ↓ HashMap$Node[].[175]
│                     ~~~~~
├─ java.util.HashMap$Node instance
│    Leaking: UNKNOWN
│    ↓ HashMap$Node.value
│                   ~~~~~
├─ java.util.ArrayList instance
│    Leaking: UNKNOWN
│    ↓ ArrayList.elementData
│                ~~~~~~~~~~~
├─ java.lang.Object[] array
│    Leaking: UNKNOWN
│    ↓ Object[].[0]
│               ~~~
├─ org.kodein.di.DIDefinition instance
│    Leaking: UNKNOWN
│    ↓ DIDefinition.binding
│                   ~~~~~~~
├─ org.kodein.di.bindings.Singleton instance
│    Leaking: UNKNOWN
│    ↓ Singleton.scope
│                ~~~~~
├─ org.kodein.di.bindings.NoScope instance
│    Leaking: UNKNOWN
│    ↓ NoScope._registry
│              ~~~~~~~~~
├─ org.kodein.di.bindings.StandardScopeRegistry instance
│    Leaking: UNKNOWN
│    ↓ StandardScopeRegistry._cache
│                            ~~~~~~
├─ java.util.concurrent.ConcurrentHashMap instance
│    Leaking: UNKNOWN
│    ↓ ConcurrentHashMap.table
│                        ~~~~~
├─ java.util.concurrent.ConcurrentHashMap$Node[] array
│    Leaking: UNKNOWN
│    ↓ ConcurrentHashMap$Node[].[0]
│                               ~~~
├─ java.util.concurrent.ConcurrentHashMap$Node instance
│    Leaking: UNKNOWN
│    ↓ ConcurrentHashMap$Node.val
│                             ~~~
├─ org.kodein.di.bindings.SingletonReference$make$1 instance
│    Leaking: UNKNOWN
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ SingletonReference$make$1.$value
│                                ~~~~~~
├─ co.my.app.DatabaseHelperImpl instance
│    Leaking: UNKNOWN
│    ↓ DatabaseHelperImpl.databaseProvider
│                               ~~~~~~~~~~~~~~~~
├─ co.my.app.Sqldelight_moduleKt$sqlDelightModule$1$1$1 instance
│    Leaking: UNKNOWN
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ Sqldelight_moduleKt$sqlDelightModule$1$1$1.$this_singleton
│                                                 ~~~~~~~~~~~~~~~
├─ org.kodein.di.bindings.NoArgBindingDIWrap instance
│    Leaking: UNKNOWN
│    ↓ NoArgBindingDIWrap._di
│                         ~~~
├─ org.kodein.di.bindings.BindingContextedDI instance
│    Leaking: UNKNOWN
│    ↓ BindingContextedDI.context
│                         ~~~~~~~
╰→ co.my.app.LauncherActivity instance
​     Leaking: YES (ObjectWatcher was watching this because co.m.app.LauncherActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
```

**Comment by SalomonBrys** [comment]
Adressed in 9e6afe1535c7d1332f9bf1364ba1e8fdc0a6296e and published as `7.1.0-master-87` in dev repository

**Comment by eygraber** [comment]
I saw the release notes mention that this might be fixed. It hasn't been (shared the trace privately; will copy it here soon).

**Comment by romainbsl** [comment]
fixed in [7.2](https://github.com/Kodein-Framework/Kodein-DI/releases/tag/7.2.0)
